File: packages/runtime-js/src/protocol.ts
import type { ValidationError } from "./types.js";
import type { ErrorSource } from "./errors.js";

export type ProjectionSpec = { [key: string]: 1 | ProjectionSpec };

export type ListOptions = {
    skip?: number;
    limit?: number;
    sort?: Record<string, 1 | -1>;
};

export type TraversalDirection = "out" | "in" | "both";
export type TraversalEmit = "nodes" | "edges" | "paths";

export type TraversalStep = {
    /** Edge schema name. */
    via: string;
    direction: TraversalDirection;
    /** Optional filter on edge properties. */
    edgeWhere?: Record<string, unknown>;
    /** Optional filter on the connected node reached via this step. */
    nodeWhere?: Record<string, unknown>;
};

export type TraversalSpec = {
    /** Starting-node schema name and where clause. Independent of the
     *  terminal-node schema, which lives on the operation. */
    start: {
        schema: string;
        where: Record<string, unknown>;
    };
    /** Explicit heterogeneous chain. Mutually exclusive with repeat/depth. */
    steps?: TraversalStep[];
    /** Homogeneous repeat — one step applied 1..N times. */
    repeat?: TraversalStep;
    depth?: { min?: number; max: number };
    /** Filter applied to terminal nodes. */
    where?: Record<string, unknown>;
    /** Output shape. Defaults to "nodes". */
    emit: TraversalEmit;
    /** Pagination/ordering applied to the emitted result set. */
    options?: ListOptions;
};

export type KeymaOperation =
    | {
          op: "list";
          schema: string;
          where?: Record<string, unknown>;
          project?: ProjectionSpec;
          options?: ListOptions;
      }
    | {
          op: "read";
          schema: string;
          where: Record<string, unknown>;
          project?: ProjectionSpec;
      }
    | {
          op: "create";
          schema: string;
          data: Record<string, unknown>;
          project?: ProjectionSpec;
      }
    | {
          op: "update";
          schema: string;
          where: Record<string, unknown>;
          data: Record<string, unknown>;
          project?: ProjectionSpec;
      }
    | {
          op: "delete";
          schema: string;
          where: Record<string, unknown>;
      }
    | {
          op: "traverse";
          /** Terminal-node schema (for hydration). */
          schema: string;
          spec: TraversalSpec;
          project?: ProjectionSpec;
      };

export type KeymaRequest = {
    operations: Record<string, KeymaOperation>;
};

export type KeymaLeafSuccess<T = unknown> = { ok: true; data: T };
export type KeymaLeafFailure = {
    ok: false;
    error: string;
    code: string;
    source: ErrorSource;
    /** Package name of the originator (plugin or adapter). Omitted for runtime errors. */
    origin?: string;
    /** Structured field errors. Present if code === "VALIDATION_FAILED". */
    errors?: ValidationError[];
    /** Free-form extras supplied by the throwing layer (e.g. {fields: [...]} for FIELD_FORBIDDEN). */
    [key: string]: unknown;
};
export type KeymaLeafResult<T = unknown> = KeymaLeafSuccess<T> | KeymaLeafFailure;

export type KeymaBatchResponse = {
    results: Record<string, KeymaLeafResult>;
};

export type Transport = (request: KeymaRequest) => Promise<KeymaBatchResponse>;

File: packages/runtime-python/src/keyma/runtime/protocol.py
"""Wire protocol — port of ``@keyma/runtime-js`` ``protocol.ts``.

Operations, requests and responses travel as plain dicts (the cross-language wire
format). The ``TypedDict`` definitions here document their shape; the runtime builds
and reads dicts directly.
"""

from __future__ import annotations

from typing import Any, Awaitable, Callable, Dict, List, Literal, Optional, TypedDict, Union

from .types import ValidationError

ProjectionSpec = Dict[str, Any]  # { field: 1 | ProjectionSpec }


class ListOptions(TypedDict, total=False):
    skip: int
    limit: int
    sort: Dict[str, int]  # field -> 1 | -1


TraversalDirection = Literal["out", "in", "both"]
TraversalEmit = Literal["nodes", "edges", "paths"]


class TraversalStep(TypedDict, total=False):
    via: str  # edge schema name
    direction: TraversalDirection
    edgeWhere: Dict[str, Any]
    nodeWhere: Dict[str, Any]


class TraversalStart(TypedDict):
    schema: str
    where: Dict[str, Any]


class TraversalDepth(TypedDict, total=False):
    min: int
    max: int


class TraversalSpec(TypedDict, total=False):
    start: TraversalStart
    steps: List[TraversalStep]
    repeat: TraversalStep
    depth: TraversalDepth
    where: Dict[str, Any]
    emit: TraversalEmit
    options: ListOptions


#: A single operation in a batch — discriminated on ``op``. Runtime value is a dict.
KeymaOperation = Dict[str, Any]


class KeymaRequest(TypedDict):
    operations: Dict[str, KeymaOperation]


class KeymaLeafSuccess(TypedDict):
    ok: Literal[True]
    data: Any


class KeymaLeafFailure(TypedDict, total=False):
    ok: Literal[False]
    error: str
    code: str
    source: str  # ErrorSource
    origin: str
    errors: List[ValidationError]


KeymaLeafResult = Dict[str, Any]  # KeymaLeafSuccess | KeymaLeafFailure


class KeymaBatchResponse(TypedDict):
    results: Dict[str, KeymaLeafResult]


#: Async transport: ``async (request) -> response``.
Transport = Callable[[KeymaRequest], Awaitable[KeymaBatchResponse]]
